Refactor Files Sources Framework for stronger typing using pydantic models

davelopez opened the pull request:
Followup to https://github.com/galaxyproject/galaxy/pull/20698#pullrequestreview-3063700318

## Update

Similar to how `User-defined FileSources` handle the configuration, this PR refactors Galaxy's FileSources framework to use **Pydantic models** for configuration management, introducing a two-tier architecture that separates template configuration from resolved configuration. This enables dynamic user-specific customization while maintaining strong type safety.

## Key Changes

### Two-Tier Configuration System

The framework now uses a dual configuration approach like in User-defined FileSources:

1. **Template Configuration** (`BaseFileSourceTemplateConfiguration`)
   - Contains raw configuration values that may include template variables
   - Supports template expansion using `${variable}` syntax
   - Example: `root: /data/${user.username}/files`

2. **Resolved Configuration** (`BaseFileSourceConfiguration`) 
   - Contains final, evaluated values after template processing
   - Used during actual file operations
   - Example: `root: /data/john_doe/files` (after template evaluation)

### Updated Base Classes

All file sources now inherit from `BaseFilesSource[TTemplateConfig, TResolvedConfig]`, providing:
- Generic typing for both configuration models
- Automatic template resolution
- Runtime context management

## Example Implementation

The Posix file source demonstrates the new pattern:

```python
class PosixTemplateConfiguration(BaseFileSourceTemplateConfiguration):
    """Template configuration with potential template variables"""
    root: Union[str, TemplateExpansion, None] = None
    # ... other template fields

class PosixConfiguration(BaseFileSourceConfiguration):
    """Resolved configuration with concrete values"""
    root: Optional[str] = None
    # ... other resolved fields

class PosixFilesSource(BaseFilesSource[PosixTemplateConfiguration, PosixConfiguration]):
    plugin_type = "posix"
    template_config_class = PosixTemplateConfiguration
    resolved_config_class = PosixConfiguration
```

File sources now declare two configuration models and assign the types to `template_config_class` and `resolved_config_class` respectively to construct and validate the configuration by the plugin loader.

When operating on the file source (listing, writing, realizing), we now have a generic `FilesSourceRuntimeContext` (replacing the previous' user_context'). Example:
```
context: FilesSourceRuntimeContext[PosixConfiguration]
```
This context already contains the resolved configuration and user data in case the operation was run by a user.


## 🎁 Bonus: File Sources Configuration Linter (with the nice help of Copilot)

This PR includes a new linting tool to validate file sources configurations. The linter helps ensure your file source configurations are syntactically correct and can be properly instantiated.

### Usage

```bash
# Validate default configuration
make files-sources-lint

# Validate with verbose output
make files-sources-lint-verbose

# Validate specific configuration file
python scripts/lint_file_sources_config.py -f path/to/file_sources_conf.yml

# Validate with custom Galaxy config
python scripts/lint_file_sources_config.py -c path/to/galaxy.yml --verbose
```

### Features

- **Colorized Output**: Clear visual feedback with colored success/error messages
- **Configuration Validation**: Attempts to instantiate each file source to catch configuration errors
- **Verbose Mode**: Shows detailed configuration being validated
- **Fail-Fast Option**: Stop on first error for quick debugging
- **Auto-Discovery**: Automatically finds Galaxy and file sources configuration files

### Example Output

`make files-sources-lint-verbose`

```
Galaxy File Sources Configuration Validator
=============================================
Using Galaxy config: ../config/galaxy.yml
Using file sources config: ../config/file_sources_conf.yml

Found 4 file source(s) to validate

[1/4] Validating file source configuration...
  Validating file source 'test-posix-source' of type 'posix'...
    Using configuration:
        type: posix
        root: /home/dlopez/sandbox/file_source_data/
        id: test-posix-source
        doc: For testing
        label: TestSource
        writable: true
    ✓ Valid: File source 'test-posix-source' (posix) configured successfully

[2/4] Validating file source configuration...
  Validating file source 'temp-source' of type 'temp'...
    Using configuration:
        type: temp
        root_path: /tmp/filesource/
        id: temp-source
        doc: For testing temp file source
        label: TEMP Source
        writable: true
    ✓ Valid: File source 'temp-source' (temp) configured successfully

[3/4] Validating file source configuration...
  Validating file source 'test-s3-source' of type 's3fs'...
    Using configuration:
        type: s3fs
        id: test-s3-source
        doc: For testing S3 file source
        label: Test S3 Source
        bucket: ${user.preferences['test-s3-source|bucket']}
        key: ${user.user_vault.read_secret('preferences/test-s3-source/access_key')}
        secret: ${user.user_vault.read_secret('preferences/test-s3-source/secret_key')}
        endpoint_url: http://localhost:9444
        writable: true
    ✓ Valid: File source 'test-s3-source' (s3fs) configured successfully

[4/4] Validating file source configuration...
  Validating file source 'zenodo' of type 'zenodo'...
    Using configuration:
        type: zenodo
        id: zenodo
        doc: Zenodo is a general-purpose open-access repository developed under the European
          OpenAIRE program and operated by CERN. It allows researchers to deposit data sets,
          research software, reports, and any other research-related digital artifacts.
        label: Zenodo
        url: https://zenodo.org
        token: ${user.user_vault.read_secret('preferences/zenodo/token')}
        public_name: ${user.preferences['zenodo|public_name']}
        writable: true
    ✓ Valid: File source 'zenodo' (zenodo) configured successfully


Validation Summary:
  Total file sources: 4
  Valid: 4
  Errors: 0

✓ All file sources validated successfully!
```


## How to test the changes?
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Refactor file source framework to use Pydantic models

For configuration management and validation

davelopez committed:
Refactor PyFilesystem2FilesSource

To improve initialization and typing

davelopez committed:
Refactor FileSourcePluginsConfig to inherit from Pydantic's BaseModel

davelopez committed:
Rename FileSourceConfiguration back to FilesSourceProperties

To avoid conflict with User templated file sources configurations for now

davelopez committed:
Adapt some types

davelopez committed:
Improve file source plugins instantiation

Create and validate their configuration before instantiating the plugin.

davelopez committed:
Refactor AnVILFilesSource

To adapt it to the new framework updates

davelopez committed:
Refactor BaseSpace file source

To adapt it to the new framework updates

davelopez committed:
Refactor Azure file source

To adapt it to the new framework updates

davelopez committed:
Refactor Dropbox file source

To adapt it to the new framework updates

davelopez committed:
Refactor FTP file source

To adapt it to the new framework updates

davelopez committed:
Refactor Galaxy stock file sources

To adapt them to the new framework updates

davelopez committed:
Refactor Posix file source

To adapt it to the new framework updates

davelopez committed:
Refactor HTTP file source

To adapt it to the new framework updates

davelopez committed:
Refactor GoogleDrive file source

To adapt it to the new framework updates

davelopez committed:
Refactor GoogleCloudStorage file source

To adapt it to the new framework updates

davelopez committed:
Refactor OneData file source

To adapt it to the new framework updates

davelopez committed:
Refactor RSpace file source

To adapt it to the new framework updates

davelopez committed:
Refactor SSH file source

To adapt it to the new framework updates

davelopez committed:
Refactor Temp file source

To adapt it to the new framework updates

davelopez committed:
Refactor WebDav file source

To adapt it to the new framework updates

davelopez committed:
Refactor S3 file source

To adapt it to the new framework updates

davelopez committed:
Refactor RDM file sources (Invenio and Dataverse)

To adapt them to the new framework updates and reduce significantly the
passing of user_context to methods.

davelopez committed:
Refactor eLabFTW file source

To adapt it to the new framework updates

davelopez committed:
Refactor DRS file source

To adapt it to the new framework updates

davelopez committed:
Refactor RemoteZip file source

To adapt it to the new framework updates

davelopez committed:
Refactor base64 file source

To adapt it to the new framework updates

davelopez committed:
Fix call to super constructor in initialization

davelopez committed:
Fix root type in PosixFileSourceConfiguration to Optional[str]

davelopez committed:
Refactor move partial_model decorator

From schema to utils.config_templates

davelopez committed:
Refactor UserData class to use context for user attributes

davelopez committed:
Add PartialFilesSourceProperties for partial updates

And update usage in fetch_drs_to_file

davelopez committed:
Fix file source type access

davelopez committed:
Refactor file source classes to remove user_context and opts parameters

All the config synchronization and variable templating happens in the base class, simplifying downstream plugin implementations.

Plugin instances can assume that they have the most up-to-date values in their config and user_data variables.

davelopez committed:
Fix linting

davelopez committed:
Add user_context handling when serializing PosixFilesSource

davelopez committed:
Rename TypeVar from T to M in partial_model decorator

To avoid conflict with existing generic in the same module

davelopez committed:
Refactor plugin instance creation logic

- Remove PluginInstance protocol

davelopez committed:
Fix type annotations in file sources

- Remove unnecessary type ignores in file source classes
- Fix some typing issues

davelopez committed:
Adapt tests to new models

davelopez committed:
Set default value for config_class in BaseFilesSource

jmchilton commented:
This is more expansive than I was imagining for a first pass but I guess we probably wanted to get here eventually right? My guess is the old dictionary properties typing we were doing was less break prone since it would keep the dictionaries intact and pass along to Pyfilesystem whereas this approach can result in swallowing properties people had been using implicitly. I'm not sure what we can do to mitigate that - is a big scary warning in the log possible if we find unexpected keys? Ideally we could also make a linter from the models and run it through the configs for main and EU.

As for sharing properties... I don't know I think that might be premature - I would think starting by sharing field definitions in galaxy.files.fields or something like that. My question in this regard is how you're handling templating. For user defined properties - there are two sets of models for each plugin - the templated version and the filled out version that gets passed along to the implementation plugin here (e.g. FtpFileSourceTemplateConfiguration is the template definition and FtpFileSourceConfiguration gets passed to the plugin). Reusing the same plugin for both... which I think is what is happening with this code is kind of abusing the type system a bit. 

```
    def _evaluate_config_props(self, user_context: "OptionalUserContext") -> dict[str, Any]:
        """Evaluate properties that may contain templated values."""
        effective_props = {}
        for key, val in self.config.model_dump(exclude_unset=True, exclude_none=True).items():
            effective_props[key] = self._evaluate_prop(val, user_context=user_context)
        return effective_props
```

I would think if both sets of plugins had both a templated and expanded set of models some of the expanded models could be reused? In the Ftp example maybe FtpFileSourceTemplateConfiguration stays with the user template stuff and FtpFileSourceTemplateConfiguration gets moved into the implementation plugin. Even then though I'm not sure - I would think the implementation plugins might want to implement more experimental features and deal with backward compatibility in different ways than the user defined plugins which are going to be more strict from the start.  

Two sets of models for each plugin type seems like a lot of work to put on plugin authors but we could generate a lot of documentation from them and build config linting tools that would make this development work more fruitful - admins are getting more from each plugin.

davelopez commented:
Thank you so much for your feedback! Really appreciate it.

> This is more expansive than I was imagining for a first pass but I guess we probably wanted to get here eventually right?

Right, I got a bit carried away 😅

> My guess is the old dictionary properties typing we were doing was less break prone since it would keep the dictionaries intact and pass along to Pyfilesystem whereas this approach can result in swallowing properties people had been using implicitly. I'm not sure what we can do to mitigate that - is a big scary warning in the log possible if we find unexpected keys?

Yeah, right now, at startup, loading file sources with unexpected keys will just fail with validation errors. We could temporarily make config models allow extra keys and detect and warn about them on load. I think the linter will also be a good idea. Do you think a "make command for linting" will be enough or something else?

> My question in this regard is how you're handling templating.

Busted 😅. I totally see what you mean now, and indeed, the templating is not properly handled in the current implementation. I will give it some thought and try to come up with a better solution, likely by using the 2-model approach you suggested. And then think about how to make user-defined and system-defined file source configurations more consistent.

In general, I think we should aim to prefer a more explicit approach to configuration, where we define the properties that are expected and required for each file source type, rather than relying on implicit behavior, but of course, we should also consider backward compatibility, as you mentioned.

Thanks again for the feedback, it's really helpful! I'll work on addressing these points and get back to you with an updated implementation. If you have any more thoughts or suggestions in the meantime, feel free to share them! 😊

davelopez committed:
Allow extra properties in PartialFilesSourceProperties

Since every file source might provide different properties.

Also fix _evaluate_config_props to avoid check templating for unnecessary cases.

davelopez committed:
Make endpoint_url optional in S3FSFileSourceConfiguration

Looks like an oversight and should have been optional according to the existing implementation.

davelopez committed:
Add unique ID to stock file source configurations

I feel like the `id` should be always required so rather than doing it optional I opted to give the stock plugins an ID.

davelopez committed:
Fix config_class declarations

To use ClassVar[type[ConfigurationType]] syntax consistently

davelopez committed:
Move file source models to module

And add generic types and utils for configuration template resolution

davelopez committed:
Refactor BaseFilesSource to support template and resolved configurations

Using generics

davelopez committed:
Adapt all PyFilesystem2FilesSources to template generics

davelopez committed:
Adapt invenio-based file sources

davelopez committed:
Fix imports for models

davelopez committed:
Adapt Base64FilesSource to use template configuration

davelopez committed:
Adapt DRS file source classes to use template configuration

davelopez committed:
Adapt eLabFTW file source to use template configuration

davelopez committed:
Adapt Posix file source to use template configuration

davelopez committed:
Adapt RemoteZipFilesSource to use template configuration

davelopez committed:
Adapt S3FS file source to use template configuration

davelopez committed:
Adapt HTTP file source to use template configuration

davelopez committed:
Adapt Galaxy stock file sources to use template configuration

davelopez committed:
Fix user_data initialization and context handling in BaseFilesSource

davelopez committed:
Update label initialization in BaseFilesSource to handle None values

davelopez committed:
Fix default values handling in file sources

davelopez committed:
Use template configuration during file source initialization

There is no point in using the resolved config util we have RuntimeContext information to be able to properly evaluate the templates.

In these cases, the previous implementation was directly using the templated value anyway because of the same reason.

davelopez committed:
Rename RuntimeContext to FilesSourceRuntimeContext

And fix user_data context handling in file sources

davelopez committed:
Fix type hints in utils for backward compatibility

davelopez committed:
Use dict instead of FilesSourceProperties in file_source_instances

davelopez committed:
Fix file source initialization by converting root LocalPath to string

As the TestPosixConfiguredFileSources expects a string as root argument as well as the pydantic model for PosixFilesSource

davelopez committed:
Add files sources configuration linter

Experimenting with Copilot

davelopez committed:
Enhance terminal output with color coding for validation messages

davelopez committed:
Remove unnecessary validation checks in file source linter

davelopez committed:
Do not use verbose mode in makefile command by default

davelopez committed:
Enhance verbose output for file source configuration in linter

davelopez committed:
Add missing FTP configuration fields to template and configuration classes

davelopez committed:
Add verbose option for file sources configuration linting in makefile

davelopez committed:
Add more optional fields in GCSFS file source configuration

davelopez committed:
Fix fetch_url_allowlist serialization in FileSourcePluginsConfig

davelopez committed:
Use template_config in places where the real config cannot be resolved.

Some functions get called before we pass any user context so we cannot access the resolved config values in those cases.

davelopez committed:
Refactor rename writeable parameter to write_intent to avoid confusion

This is to avoid unnecessary confusion with the "writable" status of a file source. The `write_intent` query parameter just signals to the framework that the user is accessing the file source with the intent of writing to it, independently of the "writable" status of the file source.

davelopez committed:
Refactor file source implementations to utilize FilesSourceRuntimeContext

I decided to go back to the previous strategy of passing the context to each operation instead of storing the context in the instance variable as this was before.

davelopez committed:
Refactor test_config_options to use template_config and resolved config

davelopez committed:
Refactor writeable parameter to write_intent in get_file_containers

Was missing in previous refactoring commits about write_intent.

davelopez committed:
Exclude defaults in template updates

So we can override with the provided defaults in _apply_defaults_to_template.

davelopez committed:
Remove some TemplateExpansion in Posix File Source

From non-user-configurable fields

davelopez committed:
Add prefer_links method and root property for backward compatibility in PosixFilesSource

davelopez committed:
Remove unhelpful docstring

davelopez committed:
Move validation script for file sources configuration

And update Makefile targets

davelopez committed:
Add entry point for galaxy file sources configuration validation

mvdbeek commented on lib/galaxy/files/sources/s3fs.py:
```diff
@@ -28,60 +26,61 @@
-    endpoint_url: int
-    user: str
-    passwd: str
-    listings_expiry_time: NotRequired[Optional[int]]
```
Was there a reason to drop this ? I think this caused https://github.com/galaxyproject/galaxy/issues/20791 ?
